Add Travis deployment

aHugues opened the pull request:
Here I modified the scripts to automatically deploy the API on the servers with Travis.

## Content

Currently they are available at 
* https://api.murder.aurelienhugues.com
* https://dev.api.murder.aurelienhugues.com

All pushes made to `master` are deployed on the main environment while all pushes to all other branches are deployed on the dev environment. 

## TODO

### Deployement script

At the moment, I use a **systemd** service for the server, auto deployed using a **bash** script on the server. 

Should I put these on the repo, or maybe add them to the documentation as their format might not be the same depending on the infrastructure...

### Frontend linking

I still need to add the API selection for the frontend, at the moment it's not setup yet.

florimondmanca commented on murders/settings/__init__.py:
```diff
@@ -1,2 +1,2 @@
 # Import default settings
-from .dev import *
+from .common import *
```
The common settings should only be use as a base settings file. It is more explicit to have production/dev settings files and choose the correct one when starting a server.
Couldn't we just keep the two different files and choose the settings module on the server? I feel like the deployment method is hindering on our development workflow, which is probably not desirable.

aHugues replied:
Yeah, but the `dev` and `prod` settings files depend strongly on the server in which to deploy. It's where the necessary environment variables are.

## Proposition 1

Maybe we should put these 2 files on the repo as default files and use other files for our purpose? 

We then would have the following structure: 
```
settings
├── __init__.py
├── common.py
├── dev.py
├── prod.py
└── server.py
```

In this case, `server.py` is not tracked in the repo and is specifically designed for the server in which the api is deployed.

## Proposition 2

We don't set needed variables in the config at all, and everything is set as environment variables in the server.

In the case of a *systemd* script, instead of 

```nginx
Environment=DJANGO_SETTINGS_MODULE=murders.settings.dev
```
we would have

```nginx
Environment=DJANGO_SETTINGS_MODULE=murders.settings.dev
Environment=ALLOWED_HOSTS=['dev.api.murder.aurelienhugues.com']
Environment=DATABASE_URL=mysql://user:password@localhost:3306/murders_dev
```

I don't know which is better. Probably **Proposition 2**?

florimondmanca replied:
I also prefer **P2** as well as in general, the more we can put in VCS the more visibility we have and the better.

I believe this is also the correct way to go — having a settings file which takes items that need to be configurable from environment variables and setting them in the server as needed. :+1:

In this case I'm not sure the `prod/dev` separation is actually needed because we can just take the minimum required amount of settings from environment.

So,

```
wsgi.py  << DJANGO_SETTINGS_MODULE=murders.settings.common
settings
├── __init__.py  << from .local import *
├── common.py << base settings, used by prod/dev servers
└── local.py  << from .common import *; DEBUG = True
```

florimondmanca replied:
Or we might as well rename `common.py` to `server.py`, which would make more sense since it's not really a "base" settings file, it's the actual settings file used in prod.

aHugues replied:
so something like `server.py` and `dev.py` or `debug.py` then?

florimondmanca replied:
I have edited my answer above.

florimondmanca commented on murders/settings/production.py:
```diff
@@ -1,3 +0,0 @@
-"""Production settings."""
-
-from .common import *
```
Why did you remove the dev settings file? Now in dev if we want to run in debug mode we have to explicitly export DEBUG=true…

aHugues replied:
I removed both config files for the deployment procedure I'm using. It might not be perfect, but we're working on it. This issue is related to the comments above

florimondmanca commented:
Also, can you confirm that keeping `ALLOWED_HOSTS = ['localhost']` is sufficient?

aHugues commented:
Concerning the `ALLOWED_HOSTS`, no it's not sufficient for it to run, but the value depends on the reverse proxy used. 

In our case it's *dev.api.aurelienhugues.com* or *api.aurelienhugues.com*. Might be able to use _*.aurelienhugues.com_ for both, but in any case it should be changed on every machine, so there's no use changing it here, it's sufficient to run the tests at least.

florimondmanca commented:
Well, `ALLOWED_HOSTS` is really a *list* of allowed hosts, so why not set it to

```
['localhost', 'dev.api.aurelinhugues.com', 'api.aurelienhugues.com']
```

directly? That way it's set for all environments we have atm (local + both servers) and no need to configure it.

aHugues commented:
yes, but in my case, you are sure that you can access the correct environment only using the correct domain. 

it adds a little security in case of a configuration error.

florimondmanca commented:
Then as you wish my fish.

aHugues commented:
I added the requested changes:

## Config

Now we have 2 config files
* `server.py` which used to be `common.py` 
* `dev.py` with the `DEBUG = True` option

We now set all needed parameters using environment variables on the service.

## Misc

I also added .vscode to the gitignore because it was pissing me off...
